lib/src/metrics/local_drain.rs

```diff
@@ -103,41 +122,131 @@ pub fn histogram_to_percentiles(hist: &Histogram<u32>) -> Percentiles {
     }
 }
 
-#[derive(Copy, Clone, Debug, PartialEq)]
-enum MetricKind {
-    Gauge,
-    Count,
-    Time,
+/// convert a collected histogram to a prometheus-compatible format
+pub fn filter_histogram(hist: &Histogram<u32>) -> FilteredMetrics {
+    let sum: u64 = hist
+        .iter_recorded()
+        .map(|item| item.value_iterated_to() * item.count_at_value() as u64)
+        .sum();
+
+    let mut count = 0;
+    let mut buckets = Vec::new();
+
+    for value in hist.iter_log(1, 2.0) {
+        count += value.count_since_last_iteration();
+
+        buckets.push(Bucket {
+            le: value.value_iterated_to(),
+            count,
+        });
+    }
+
+    FilteredMetrics {
+        inner: Some(filtered_metrics::Inner::Histogram(FilteredHistogram {
+            sum,
+            count,
+            buckets,
+        })),
+    }
+}
+
+/// a map of metric_name -> metric value
+#[derive(Debug, Clone)]
+pub struct MetricsMap {
+    map: BTreeMap<String, AggregatedMetric>,
 }
 
-#[derive(Clone, Debug, PartialEq)]
-enum MetricMeta {
-    Cluster,
-    ClusterBackend,
+impl MetricsMap {
+    fn new() -> Self {
+        Self {
+            map: BTreeMap::new(),
+        }
+    }
+
+    /// convert a metrics map to a map of filtered metrics,
+    /// perform a double conversion for time metrics: to percentiles and to histogram
+    fn to_filtered_metrics(&self, filter_by_names: &[String]) -> BTreeMap<String, FilteredMetrics> {
```

Comment:
this function might not be very efficient.

`self.map.clone()` is expensive. Maybe use a `filter` instead of `retain`?

Maybe the `insert` in a loop could also be replaced by a `collect`.

Reply:
I rewrote the function in a functionnal way, with `filter` and `collect` ;-)